### PR TITLE
Add auto-attached card

### PR DESCRIPTION
**Pull request by FadhlanR**
Ticket: CS-6421

- [x] top-most card in stack(s) is/are shown in Chat Input as "auto-attached"
- [x] "auto-attached" cards show with a dashed border
- [x] "auto-attached" cards show a tooltip on hover: "Topmost card is shared automatically"
- [x] An "X" is shown at right side of pill and can be used to remove the auto-attached card so that it is not attached when the message is sent
- [x] When the top-most card changes (i.e. the user navigates), any current auto-attached cards are removed and the new top-most card(s) is/are attached
- [x] If a prospective auto-attached card is already explicitly attached by the user via the Attach Card button, it does not appear as auto-attached

https://github.com/cardstack/boxel/assets/12637010/f236eaaa-6979-4594-b145-ae200a503ce1

### CODE REVIEW

File: packages/host/app/components/ai-assistant/card-picker/index.gts
Comment (jurgenwerk) [inline packages/host/app/components/ai-assistant/card-picker/index.gts]: Is this condition needed?

File: packages/host/app/components/matrix/room-input.gts
Comment (jurgenwerk) [inline packages/host/app/components/matrix/room-input.gts]:
```suggestion
    // If card doesn't exist in `cardsToAttach`,
```

**Review by jurgenwerk — APPROVED** [review]

**Comment by jurgenwerk** [comment]
1. I noticed a bug in attached cards, nothing happens when I click on X. Not sure if introduced by this PR.

image

2. When I remove the autoattached card, and navigate to other cards in the stack, should the new topmost card now be autoattached? I am not seeing this behaviour here but wondering if it should be like that

2. Does "◻︎ Allow access to the cards you can see at the top of your stacks" still make sense now that we are autoattaching them?

**Comment by IanCal** [comment]
This should replace the "share the topmost card" functionality, it'll be part of the context, so we should remove that checkbox.

This updates the message type so that auto-attached and attached cards can both coexist -> https://github.com/cardstack/boxel/pull/1045

Potentially we could change the structure here so that they all sit in "attached" cards and then the concept of an open card is just the ID of what the user is looking at? 

Should we block the index card from being attached?

**Comment by FadhlanR** [comment]
> 1. I noticed a bug in attached cards, nothing happens when I click on X. Not sure if introduced by this PR.
> 
> image
> 2. When I remove the autoattached card, and navigate to other cards in the stack, should the new topmost card now be autoattached? I am not seeing this behaviour here but wondering if it should be like that
> 3. Does "◻︎ Allow access to the cards you can see at the top of your stacks" still make sense now that we are autoattaching them?

@jurgenwerk Nice catch! I made a few updates.

- I fixed the removing attached card issue.
- I implemented the behavior when user navigating to another card after removing the auto-attached card. Now it must displays auto-attached card if user navigating to different card.
- I removed the "share the topmost card" checkbox.


https://github.com/cardstack/boxel/assets/12637010/c0a8ad56-ed1f-45bc-abdd-3d68d87d37c3

**Comment by FadhlanR** [comment]
> This should replace the "share the topmost card" functionality, it'll be part of the context, so we should remove that checkbox.
> 
> This updates the message type so that auto-attached and attached cards can both coexist -> #1045
> 
> Potentially we could change the structure here so that they all sit in "attached" cards and then the concept of an open card is just the ID of what the user is looking at?
> 
> Should we block the index card from being attached?

It make sense to block the index card from being attached, I'll make a change for that.

Yeah I agree, it seems I have to change the structure here because currently, we can only choose between sending attached cards or sending a message with context for open card and it breaks lots of tests. I think I have to wait until your PR get merged, what do you think @IanCal ?

**Comment by IanCal** [comment]
@FadhlanR I think my PR should be passing the linting now, I don't mind which way we do this, if it makes sense we could merge mine then I can take over this branch and update it to match the format, then change openCards to just have the ID?

**Comment by FadhlanR** [comment]
> @FadhlanR I think my PR should be passing the linting now, I don't mind which way we do this, if it makes sense we could merge mine then I can take over this branch and update it to match the format, then change openCards to just have the ID?

It sounds good, you can do that. Thanks!

**Review by burieberry — APPROVED** [review]
I updated the tests yesterday. Do you want me to fix the conflict on your branch?

File: packages/host/tests/integration/components/operator-mode-test.gts
Comment (burieberry) [inline packages/host/tests/integration/components/operator-mode-test.gts]: This will still work but since now a new session is created automatically, you can replace these with a single line `let roomId = await openAiAssistant();` and then use `roomId` when making assertions instead of using `test_a`. 

The other change is I put the matrix tests inside a module because it was frustrating to find them.

Comment (IanCal) [reply packages/host/tests/integration/components/operator-mode-test.gts]: great thanks!

**Comment by habdelra** [comment]
I noticed that the failing matrix tests are recently added tests. I just double checked and these are passing for me in main. I wonder if the upstream merge worked properly?

**Comment by burieberry** [comment]
> I noticed that the failing matrix tests are recently added tests. I just double checked and these are passing for me in main. I wonder if the upstream merge worked properly?

I think I see where it's failing. Since in this PR there's an autoattached card, there are 2 cards instead of 1 card. There's a `removeAutoAttachedCard` helper to get around that. I can update. 

I think ideally we shouldn't be auto-attaching the index card, but I don't know what the decision was about this.

**Comment by habdelra** [comment]
> > I noticed that the failing matrix tests are recently added tests. I just double checked and these are passing for me in main. I wonder if the upstream merge worked properly?
> 
> I think I see where it's failing. Since in this PR there's an autoattached card, there are 2 cards instead of 1 card. There's a `removeAutoAttachedCard` helper to get around that. I can update.
> 
> I think ideally we shouldn't be auto-attaching the index card, but I don't know what the decision was about this.

agreed. it's just blindly including the top most card including if the index card is the top most card. i don't think that was an actual decision anyone made. i think it's a bug to be honest.

**Comment by IanCal** [comment]
I'll raise a new ticket and pr for ignoring the index card

**Pull request merged**
